### PR TITLE
Set default linker to lld for Amazon Linux 2023

### PR DESCRIPTION
**Pull request by futurejones**
The `gold` linker is no longer available in Amazon Linux 2023 and requires `lld` to be set as the default linker.

This removes the need to build the gold linker from source as proposed here - https://github.com/apple/swift-installer-scripts/pull/278

These changes only effect `amazonlinux 2023` and should not impact on the current builds of `amazonlinux 2` .

### CODE REVIEW

**Comment by futurejones** [comment]
@sebsto, can you look at this and give some feedback please.

File: lib/Driver/UnixToolChains.cpp
Comment (finagolfin) [inline lib/Driver/UnixToolChains.cpp:116]: Got a better way to organize this: can you [setup an AmazonLinux toolchain driver instead, like OpenBSD did](https://github.com/apple/swift/blob/ef8825b/lib/Driver/ToolChains.h#L198), [instantiate that based on the vendor](https://github.com/apple/swift/blob/ef8825b/lib/Driver/Driver.cpp#L356), then [add whatever logic like this that you need in a `getDefaultLinker()` override instead](https://github.com/apple/swift/blob/bf137cb30dda404ffc6b15658e61a4f41187d2cb/lib/Driver/UnixToolChains.cpp#L460)? That would be cleaner.

**Comment by finagolfin** [comment]
CMake approach seems fine if it works, but note that you will need a `swift-driver` pull also.

**Comment by futurejones** [comment]
@finagolfin The `swift-driver` issue has already been fixed with https://github.com/apple/swift-driver/pull/1545

File: lib/Driver/UnixToolChains.cpp
Comment (futurejones) [reply lib/Driver/UnixToolChains.cpp:116]: I will take a look at your suggestion, but as the move is to make `lld` the default linker hopefully none of this code will be needed soon.

Comment (finagolfin) [reply lib/Driver/UnixToolChains.cpp:116]: Sure, if the Swift devs say they will be removing `lib/Driver/` altogether and don't need it changed too, just the CMake change should be enough.

**Comment by sebsto** [comment]
> @sebsto, can you look at this and give some feedback please.

Thank you for working on this ! The maintainers of this project will be more qualified than me to suggest the correct approach :-) 
As there is no decision taken to use `lld` as default linker, I would favor [the approach suggested by finagolfin](https://github.com/apple/swift/pull/72049#discussion_r1510205563) as it seems a better longer term decision.

File: lib/Driver/UnixToolChains.cpp
Comment (futurejones) [reply lib/Driver/UnixToolChains.cpp:116]: I have had a look at this and I am not sure how this could work in the case of Amazon Linux.
With `OpenBSB` we are dealing with a specific `OSType` where as `Amazon Linux` is just one of the many `Linux` distributions. 
There is no `VendorType`, `OSType` or `EnvironmentType` that can differentiate between different Linux distributions.
As we don't want to effect any other of the Linux distributions I think the code I have put forward is the simplest way to achieve this.

Comment (finagolfin) [reply lib/Driver/UnixToolChains.cpp:116]: > There is no VendorType, OSType or EnvironmentType that can differentiate between different Linux distributions.

You are right: I assumed there would be, like [there is one for `SUSE`, but there isn't one for Amazon](https://github.com/apple/llvm-project/blob/stable/20230725/llvm/include/llvm/TargetParser/Triple.h#L166).

> As we don't want to effect any other of the Linux distributions I think the code I have put forward is the simplest way to achieve this.

I don't think this is interfering as little as it should though. I suggest you spin this logic off as a separate internal method `bool isAmazonLinux2023Host()`, then simply modify the Android line below: `if (getTriple().isAndroid() || isAmazonLinux2023Host())`. That way, at least nobody has to read all this logic if they don't care about AL2023.

Comment (futurejones) [reply lib/Driver/UnixToolChains.cpp:116]: That sounds like a good idea, I will make those changes.

File: lib/Driver/UnixToolChains.cpp
Comment (etcwilde) [inline lib/Driver/UnixToolChains.cpp:12]: If possible, please use the LLVM APIs available. We're already including them.

**Review by etcwilde — COMMENTED** [review]
Don't love it, but we're expecting to delete it so I'll be less picky. I would still get @artemcm's approval though. I'll land https://github.com/apple/swift-driver/pull/1545 once we know what our story is with `lld` and the removal of the start/stop symbols. I definitely prefer one of @al45tair (https://github.com/apple/swift/pull/71373#issuecomment-1971672506), @keith (https://github.com/apple/swift/pull/67476), or @kateinoigakukun's (https://github.com/apple/swift/pull/71373) approach over mine in https://github.com/apple/llvm-project/pull/8214.

**Comment by al45tair** [comment]
> I definitely prefer one of @al45tair ([#71373 (comment)](https://github.com/apple/swift/pull/71373#issuecomment-1971672506)), @keith (#67476), or @kateinoigakukun's (#71373) approach over mine in [apple/llvm-project#8214](https://github.com/apple/llvm-project/pull/8214).

FWIW, I'm working on a more complete fix for that at the moment.

File: lib/Driver/UnixToolChains.cpp
Comment (futurejones) [reply lib/Driver/UnixToolChains.cpp]: @etcwilde, I have removed `#include <iostream>` as it was only being used for print some log messages while I was debugging.

**Comment by futurejones** [comment]
@al45tair @artemcm @sebsto @etcwilde @tkremenek 
Is this moving forward? Can I get some indication as to whether this is going to get approved and merged or has it been abandoned in favor of another option?
Thanks.

**Comment by etcwilde** [comment]
I'm waiting on Alastair's patch that fixes the issue where lld deletes our runtime type info start/stop sections. That's blocked on a bug in lldb not being able to handle multiple sections with the same name but different flags in the same binary.

**Comment by etcwilde** [comment]
Sorry, I was thinking of the other PR for changing how the driver selects the linker.

I'm definitely not a huge fan of checking the distro names, that doesn't scale and seems brittle. It's for the old driver and I don't have a better idea though. Given that this is @artemcm's world, I defer final vote to him. Alternatively, @al45tair might have an idea for checking distribution info that doesn't involve hardcoding names into the driver.

**Comment by al45tair** [comment]
@swift-ci Please smoke test Windows platform

**Review by artemcm — APPROVED** [review]

File: CMakeLists.txt
Comment (drodriguez) [inline CMakeLists.txt:953]: This should be reverted.

The minimum CMake version is 3.19.6 (says so at the top of the file).

This key only appeared on CMake 3.22.

**Comment by finagolfin** [comment]
I think we lowered our standards, as stated above, because the real fix is in apple/swift-driver#1545 and this legacy Driver is slated to be deleted soon.

**Comment by etcwilde** [comment]
We still need something that is valid for the stated version of CMake. I'm not personally opposed to updating the CMake requirements, but that should come from an explicit decision to do so.

**Comment by finagolfin** [comment]
@futurejones, would you submit a revision to the CMake change that will work with older CMake also?

File: CMakeLists.txt
Comment (futurejones) [reply CMakeLists.txt:953]: The minimum cmake version required for the swift toolchain build has been `"cmake": "v3.24.2"` since version 5.10

https://github.com/apple/swift/blob/main/utils/update_checkout/update-checkout-config.json#L232

https://github.com/apple/swift/pull/67018

Comment (drodriguez) [reply CMakeLists.txt:953]: That's the version used by the `build-script` for CI, and only on Linux and BSD. The CMake files should be compatible with the version required by the `cmake_minimum_required` instruction. This is failing for a macOS which a 3.20 version of CMake (enough to build LLVM and Swift using only CMake).

Comment (futurejones) [reply CMakeLists.txt:953]: I think the best option here is to update the minimum cmake version to match the version used in the CI. 
This was supposed to have been done already and it makes no sense to be using a cmake version as old as v3.19.6

Comment (finagolfin) [reply CMakeLists.txt:953]: I think @etcwilde plans to significantly increase the minimum CMake version this year anyway, so it would make sense to increase it to 3.22 now itself.

Evan, wdyt?

Comment (drodriguez) [reply CMakeLists.txt:953]: But things need to be done in certain order, and not as a reaction to introducing a dependency to check for a specific Linux distribution, and have a broken `CMakeLists.txt` at the root of the project.

**Comment by al45tair** [comment]
> **drodriguez**:
> I think the feelings above about hardcoding distribution names were correct.

> **finagolfin**:
> I think we lowered our standards, as stated above, because the real fix is in https://github.com/apple/swift-driver/pull/1545 and this legacy Driver is slated to be deleted soon.

Exactly this. Hardcoding distribution names isn't great, but it lets us make forward progress and we have a better fix in the works already.

The CMake version problem is a bigger issue, mind. That said, I think we could just do

```cmake
if(CMAKE_VERSION VERSION_GREATER "3.22")
cmake_host_system_information(RESULT DISTRO_NAME  QUERY DISTRIB_PRETTY_NAME)
endif()
```

and tell people using Amazon Linux 2023 that they need at least CMake 3.22 for things to work. @drodriguez since you have an old CMake installed, could you confirm that that change fixes things for you? If it does, someone can raise a PR with that change in it to make the actual fix.

> **drodriguez**:
> This could have been a configuration in a cache file used to build for that particular distribution.

If we're talking about official distributions of Swift, either provided by swift.org or by the Amazon Linux folks, sure. I'm not sure that's such a great solution in general though — it'd mean having to have special instructions for building on Amazon Linux 2023 and above. And we already have a better solution in https://github.com/apple/swift-driver/pull/1545.

File: CMakeLists.txt
Comment (al45tair) [reply CMakeLists.txt:953]: Assuming #73001 fixes the problem for now, we can discuss the CMake version separately.

**Comment by futurejones** [comment]
> having to have special instructions for building on Amazon Linux 2023

There should be no need for any special instructions for Amazon Linux 2023 as the default install version of CMake is already `v3.22.2`

**Comment by sebsto** [comment]
@futurejones @al45tair I managed to build Swift 6 on ALI2023 thanks to your PRs.

The trick is to install `ld.gold` and Swift 5.10 on the build machine to create the tarball.
See the script here https://gist.github.com/sebsto/6409ed9dbf4c9dd1488e049165f751aa

However, when I try to use the package I created on a new ALI2023 machine, the swift driver (or compiler) still expects to find `ld.gold`
I thought this PR would cause the Swift 6.x driver to pickup lld by default. See below.

```
[ec2-user@ip-172-31-25-1 test]$ swift --version 
Swift version 6.0.1-dev (LLVM 73500bf55acff5f, Swift 1e770fa56a3fff2)
Target: x86_64-unknown-linux-gnu

[ec2-user@ip-172-31-25-1 test]$ swift package init --type executable 
Creating executable package: test
Creating Package.swift
Creating .gitignore
Creating Sources/
Creating Sources/main.swift

[ec2-user@ip-172-31-25-1 test]$ swift build
error: 'test': Invalid manifest (compiled with: ["/home/ec2-user/swift/usr/bin/swiftc", "-vfsoverlay", "/tmp/TemporaryDirectory.tlr5uA/vfs.yaml", "-L", "/home/ec2-user/swift/usr/lib/swift/pm/ManifestAPI", "-lPackageDescription", "-Xlinker", "-rpath", "-Xlinker", "/home/ec2-user/swift/usr/lib/swift/pm/ManifestAPI", "-swift-version", "6", "-I", "/home/ec2-user/swift/usr/lib/swift/pm/ManifestAPI", "-package-description-version", "6.0.0", "/home/ec2-user/test/Package.swift", "-o", "/tmp/TemporaryDirectory.6Jg8Bw/test-manifest"])
error: link command failed with exit code 1 (use -v to see invocation)
clang: error: invalid linker name in argument '-fuse-ld=gold'
error: 'test': Invalid manifest (compiled with: ["/home/ec2-user/swift/usr/bin/swiftc", "-vfsoverlay", "/tmp/TemporaryDirectory.4iTMeI/vfs.yaml", "-L", "/home/ec2-user/swift/usr/lib/swift/pm/ManifestAPI", "-lPackageDescription", "-Xlinker", "-rpath", "-Xlinker", "/home/ec2-user/swift/usr/lib/swift/pm/ManifestAPI", "-swift-version", "6", "-I", "/home/ec2-user/swift/usr/lib/swift/pm/ManifestAPI", "-package-description-version", "6.0.0", "/home/ec2-user/test/Package.swift", "-o", "/tmp/TemporaryDirectory.CW8Y5W/test-manifest"])
error: link command failed with exit code 1 (use -v to see invocation)
clang: error: invalid linker name in argument '-fuse-ld=gold'
```

What did I do wrong ?

**Comment by finagolfin** [comment]
> Is the change not merged yet on the 6.x branches ? (Just found out https://github.com/swiftlang/swift-driver/pull/1595)

As I noted six months ago, `swift-driver` is where this is really set. If you pull in that linked pull, it should work.

**Comment by sebsto** [comment]
@finagolfin is this the one you refer too ?
https://github.com/swiftlang/swift-driver/pull/1545

This has been merged into main but is not in 6.0.1 nor 6.0.2 :-(

**Comment by finagolfin** [comment]
Yes, it was submitted to 6.0 a while back in

**Pull request merged**
